File: apps/api/app/routers/orchestrator.py
"""Orchestrator API: autonomous task scheduling and execution endpoints.

Endpoints:
  POST /orchestrator/pulse     — Evaluate schedules, create & optionally execute tasks
  POST /orchestrator/trigger   — Manually trigger a specific schedule
  POST /orchestrator/execute/{task_id} — Execute a specific orchestrator task
  GET  /orchestrator/status    — Schedules, active tasks, recent history
  GET  /orchestrator/schedules — List schedule definitions

Note: pulse/trigger/execute are declared as sync `def` (not `async def`) so
FastAPI runs them in a thread pool, preventing event-loop starvation during
long-running LLM pipelines.
"""

import logging
import re

from fastapi import APIRouter, Depends, HTTPException, Path

from app.auth import get_current_user, CurrentUser
from app.schemas.orchestrator import (
    PulseRequest,
    PulseResult,
    TriggerRequest,
    TriggerResult,
    ExecuteResult,
    OrchestratorStatus,
)
from app.services.agent_orchestrator import (
    pulse,
    execute_task,
    trigger_schedule,
    get_status,
    SCHEDULES,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/orchestrator", tags=["orchestrator"])

_TASK_ID_RE = re.compile(r"^[A-Za-z0-9_-]{1,80}$")


@router.post("/pulse", response_model=PulseResult)
def orchestrator_pulse(
    body: PulseRequest,
    user: CurrentUser = Depends(get_current_user),
):
    """Run the orchestrator pulse: check schedules, create & execute due tasks."""
    return pulse(user.id, auto_execute=body.auto_execute, force=body.force)


@router.post("/trigger", response_model=TriggerResult)
def orchestrator_trigger(
    body: TriggerRequest,
    user: CurrentUser = Depends(get_current_user),
):
    """Manually trigger a specific schedule, ignoring cooldown."""
    try:
        return trigger_schedule(
            user.id, body.schedule_id, auto_execute=body.auto_execute,
        )
    except ValueError:
        raise HTTPException(400, "Invalid schedule or no active brand")


@router.post("/execute/{task_id}", response_model=ExecuteResult)
def orchestrator_execute(
    task_id: str = Path(..., max_length=80),
    user: CurrentUser = Depends(get_current_user),
):
    """Execute a specific orchestrator task."""
    if not _TASK_ID_RE.match(task_id):
        raise HTTPException(400, "Invalid task ID format")
    try:
        result = execute_task(task_id, user.id)
        return ExecuteResult(
            status=result.get("status", "unknown"),
            task_id=task_id,
            deliverable_id=result.get("deliverable_id"),
            error=result.get("error"),
            details=result,
        )
    except ValueError:
        raise HTTPException(400, "Task not found or cannot be executed")


@router.get("/status", response_model=OrchestratorStatus)
def orchestrator_status(
    user: CurrentUser = Depends(get_current_user),
):
    """Get orchestrator status: schedules, active tasks, recent history."""
    return get_status(user.id)


@router.get("/schedules")
def list_schedules(user: CurrentUser = Depends(get_current_user)):
    """List all schedule definitions."""
    return {
        "schedules": [
            {
                "id": s["id"],
                "name": s["name"],
                "agent_id": s["agent_id"],
                "task_type": s["task_type"],
                "priority": s["priority"],
                "day_of_week": s["day_of_week"],
                "hour": s["hour"],
            }
            for s in SCHEDULES
        ]
    }

File: apps/web/src/lib/api/orchestrator.ts
/**
 * Orchestrator API client — autonomous task scheduling and execution.
 */
import { apiFetch } from "./client";

// ── Types ─────────────────────────────────────────────────

export interface ScheduleDefinition {
  id: string;
  name: string;
  agent_id: string;
  task_type: string;
  priority: string;
  day_of_week: number;
  hour: number;
}

export interface ScheduleState {
  id: string;
  name: string;
  agent_id: string;
  task_type: string;
  is_due: boolean;
  has_recent_run: boolean;
  last_run: {
    id: string;
    status: string;
    created_at: string;
    completed_at: string | null;
  } | null;
}

export interface PulseResult {
  timestamp: string;
  created_tasks: Record<string, unknown>[];
  skipped: { schedule_id: string; reason: string }[];
  executed: Record<string, unknown>[];
  active_brand: { id: string; name: string } | null;
}

export interface TriggerResult {
  task: Record<string, unknown>;
  execution: Record<string, unknown> | null;
}

export interface OrchestratorExecuteResult {
  status: string;
  task_id: string;
  deliverable_id: string | null;
  error: string | null;
  details: Record<string, unknown>;
}

export interface OrchestratorStatus {
  timestamp: string;
  schedules: ScheduleState[];
  active_tasks: Record<string, unknown>[];
  recent_completed: Record<string, unknown>[];
}

// ── API Methods ───────────────────────────────────────────

export const orchestratorApi = {
  /** Run the orchestrator pulse: check schedules, create & execute due tasks. */
  pulse: (opts: { auto_execute?: boolean; force?: boolean } = {}) =>
    apiFetch<PulseResult>("/orchestrator/pulse", {
      method: "POST",
      body: JSON.stringify({
        auto_execute: opts.auto_execute ?? false,
        force: opts.force ?? false,
      }),
    }),

  /** Manually trigger a specific schedule, ignoring cooldown. */
  trigger: (scheduleId: string, autoExecute: boolean = true) =>
    apiFetch<TriggerResult>("/orchestrator/trigger", {
      method: "POST",
      body: JSON.stringify({
        schedule_id: scheduleId,
        auto_execute: autoExecute,
      }),
    }),

  /** Execute a specific orchestrator task. */
  execute: (taskId: string) =>
    apiFetch<OrchestratorExecuteResult>(`/orchestrator/execute/${taskId}`, {
      method: "POST",
    }),

  /** Get orchestrator status: schedules, active tasks, recent history. */
  status: () => apiFetch<OrchestratorStatus>("/orchestrator/status"),

  /** List all schedule definitions. */
  schedules: () =>
    apiFetch<{ schedules: ScheduleDefinition[] }>("/orchestrator/schedules"),
};
